lib/ecto/type.ex

```diff
@@ -263,39 +264,64 @@ defmodule Ecto.Type do
   to be a valid Ecto type, as it will be sent to the
   underlying data store.
 
-      iex> dump(:string, nil)
+      iex> dump(:string, nil, %{})
       {:ok, %Ecto.Query.Tagged{value: nil, type: :string}}
-      iex> dump(:string, "foo")
+      iex> dump(:string, "foo", %{})
       {:ok, "foo"}
 
-      iex> dump(:integer, 1)
+      iex> dump(:integer, 1, %{})
       {:ok, 1}
-      iex> dump(:integer, "10")
+      iex> dump(:integer, "10", %{})
       :error
 
-      iex> dump(:binary, "foo")
+      iex> dump(:binary, "foo", %{})
       {:ok, %Ecto.Query.Tagged{value: "foo", type: :binary}}
-      iex> dump(:binary, 1)
+      iex> dump(:binary, 1, %{})
       :error
 
-      iex> dump({:array, :integer}, [1, 2, 3])
+      iex> dump({:array, :integer}, [1, 2, 3], %{})
       {:ok, [1, 2, 3]}
-      iex> dump({:array, :integer}, [1, "2", 3])
+      iex> dump({:array, :integer}, [1, "2", 3], %{})
       :error
-      iex> dump({:array, :binary}, ["1", "2", "3"])
+      iex> dump({:array, :binary}, ["1", "2", "3"], %{})
       {:ok, %Ecto.Query.Tagged{value: ["1", "2", "3"], type: {:array, :binary}}}
 
+      iex> dump(:binary_id, "7d5bed50-e8ec-4a74-b863-eb977f3db92e", %{binary_id: Ecto.UUID})
+      {:ok, %Ecto.Query.Tagged{tag: nil, type: :uuid,
+        value: <<125, 91, 237, 80, 232, 236, 74, 116, 184, 99, 235, 151, 127, 61, 185, 46>>}}
+
   """
-  @spec dump(t, term) :: {:ok, term} | :error
-  def dump(type, nil) do
+  @spec dump(t, term, map) :: {:ok, term} | :error
+  def dump(type, nil, _id_types) do
```

Comment:
We need to catch nils here because we don't have id_types when schemas are compiles, and they check the default value by dumping it. I think it won't work with other values though...